python/lsst/daf/butler/core/_topology.py

```diff
@@ -171,25 +179,156 @@ def temporal(self) -> Optional[TopologicalFamily]:
     database-storage representation, and whose instances act like a
     SQLAlchemy-based column expression.
     """
 
+    NAME: ClassVar[str]
+    """Name to use for this logical column in tables (`str`).
+
+    If the representation actually uses multiple columns, this will just be
+    part of the names of those columns.  Queries (and tables that represent
+    materialized queries) may use a different name (via the ``name`` parameters
+    to various methods) in order to disambiguate between the regions associated
+    with different tables.
+    """
+
+    SPACE: ClassVar[TopologicalSpace]
+    """Topological space in which the regions represented by this class exist.
+    """
+
+    @classmethod
+    @abstractmethod
+    def makeFieldSpecs(cls, nullable: bool, name: Optional[str] = None, **kwargs: Any
+                       ) -> Tuple[ddl.FieldSpec, ...]:
+        """Make one or more `ddl.FieldSpec` objects that reflect the fields
+        that must be added to a table for this representation.
+
+        Parameters
+        ----------
+        nullable : `bool`
+            If `True`, the timespan is permitted to be logically ``NULL``
+            (mapped to `None` in Python), though the correspoding value(s) in
+            the database are implementation-defined.  Nullable timespan fields
+            default to NULL, while others default to (-∞, ∞).
+        name : `str`, optional
+            Name for the logical column; a part of the name for multi-column
+            representations.  Defaults to ``cls.NAME``.
+        **kwargs
+            Keyword arguments are forwarded to the `ddl.FieldSpec` constructor
+            for all fields; implementations only provide the ``name``,
+            ``dtype``, and ``default`` arguments themselves.
+
+        Returns
+        -------
+        specs : `tuple` [ `ddl.FieldSpec` ]
+            Field specification objects; length of the tuple is
+            subclass-dependent, but is guaranteed to match the length of the
+            return values of `getFieldNames` and `update`.
+        """
+        raise NotImplementedError()
+
     @classmethod
     @abstractmethod
-    def fromSelectable(cls: Type[_S], selectable: sqlalchemy.sql.FromClause) -> _S:
+    def getFieldNames(cls, name: Optional[str] = None) -> Tuple[str, ...]:
+        """Return the actual field names used by this representation.
+
+        Parameters
+        ----------
+        name : `str`, optional
+            Name for the logical column; a part of the name for multi-column
+            representations.  Defaults to ``cls.NAME``.
+
+        Returns
+        -------
+        names : `tuple` [ `str` ]
+            Field name(s).  Guaranteed to be the same as the names of the field
+            specifications returned by `makeFieldSpecs`.
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    @abstractmethod
+    def update(cls, extent: Optional[_R], *, name: Optional[str] = None,
```

Comment:
Not really; I tend to just start forcing some kwargs after the total number starts to get large, because I find it very hard to read calls with too many positional args.  But I was probably overzealous here, and I'll just make these all positional.